File: scatter/inst/htmlwidgets/scatter.js
// NOTE http://www.htmlwidgets.org/develop_intro.html

HTMLWidgets.widget({

    name: 'scatter',

    type: 'output',

    factory: function (el, width, height) {

        var chart = new d3.Scatter(el.id);

        // booleans converted from R may be true, false, 'TRUE', or 'FALSE'
        function getBool (bool) {
            return (bool === true || bool === 'TRUE');
        }

        return {
            renderValue: function (x) {
                var xKey = x.options.xKey;
                var yKey = x.options.yKey;
                var rKey = x.options.rKey;
                var fKeyCategorical = x.options.fKeyCategorical;
                var fKeyContinuous = x.options.fKeyContinuous;
                var categorical = getBool(x.options.categorical);
                var loColor = x.options.loColor;
                var mdColor = x.options.mdColor;
                var hiColor = x.options.hiColor;
                var numColors = x.options.numColors;
                var noTransition = getBool(x.options.noTransition);
                var hardReload = getBool(x.options.hardReload);
                var newData = getBool(x.options.newData);

                if (!chart.data || hardReload) {
                    var data = HTMLWidgets.dataframeToD3(x.data);
                    var options = {
                        width: width,
                        height: height,
                        rKey: rKey,
                        fKeyCategorical: fKeyCategorical,
                        fKeyContinuous: fKeyContinuous,
                        categorical: categorical,
                        loColor: loColor,
                        hiColor: hiColor,
                        noTransition: noTransition
                    };

                    chart.initialize(data, xKey, yKey, options);

                } else if (newData) {
                    var data = HTMLWidgets.dataframeToD3(x.data);
                    var options = {
                        rKey: rKey,
                        fKeyCategorical: fKeyCategorical,
                        fKeyContinuous: fKeyContinuous,
                        noTransition: noTransition
                    };

                    chart.updateData(data, xKey, yKey, options);

                } else if (xKey !== chart.xKey) {

                    chart.updateXKey(xKey);

                } else if (yKey !== chart.yKey) {

                    chart.updateYKey(yKey);

                } else if (rKey !== chart.rKey) {

                    chart.updateRKey(rKey);

                } else if (fKeyCategorical !== chart.fKeyCategorical) {

                    chart.updateFKeyCategorical(fKeyCategorical);

                } else if (fKeyContinuous !== chart.fKeyContinuous) {

                    chart.updateFKeyContinuous(fKeyContinuous);

                } else if (categorical !== chart.categorical) {

                    chart.updateColorScaling(categorical);

                } else if (
                    loColor !== chart.loColor ||
                    mdColor !== chart.mdColor ||
                    hiColor !== chart.hiColor ||
                    numColors !== chart.numColors) {

                    chart.updateColors(loColor, mdColor, hiColor, numColors);

                }
            },

            resize: function (width, height) {
                chart.resize(width, height);
            }
        };
    }
});

File: scatter/R/scatter.R
#' @import htmlwidgets
#' @export

# You can learn more about package authoring with RStudio at:
#
#   http://r-pkgs.had.co.nz/
#
# Some useful keyboard shortcuts for package authoring:
#
#   Build and Reload Package:  'Cmd + Shift + B'
#   Check Package:             'Cmd + Shift + E'
#   Test Package:              'Cmd + Shift + T'
scatter <- function (matrix,
                     width = NULL,
                     height = NULL,
                     xKey,
                     yKey,
                     rKey,
                     fKeyCategorical,
                     fKeyContinuous,
                     categorical = TRUE,
                     loColor = '#3366cc',
                     mdColor = 'darkgrey',
                     hiColor = '#109618',
                     numColors = 256,
                     enableTransitions = TRUE,
                     hardReload = FALSE,
                     newData = FALSE) {

    # read the matrix
    data <- matrix
    origColnames <- colnames(data)
    data <- cbind(rownames(data), data)
    colnames(data) <- c('key', origColnames)

    options <- list(
        id = 'scatter',
        xKey = xKey,
        yKey = yKey,
        rKey = rKey,
        fKeyCategorical = fKeyCategorical,
        fKeyContinuous = fKeyContinuous,
        categorical = categorical,
        loColor = loColor,
        mdColor = mdColor,
        hiColor = hiColor,
        numColors = numColors,
        enableTransitions = enableTransitions,
        hardReload = hardReload,
        newData = newData
    )

    # pass the data and settings using 'x'
    x <- list(
        data = data,
        options = options
    )

    htmlwidgets::createWidget('scatter', x, width = width, height = height)
}

#' @export
scatterOutput <- function (outputId, width = '100%', height = '400px') {
    shinyWidgetOutput(outputId, 'scatter', width, height, package = 'scatter')
}

#' @export
renderScatter <- function (expr, env = parent.frame(), quoted = FALSE) {
    if (!quoted) { expr <- substitute(expr) } # force quoted
    shinyRenderWidget(expr, scatterOutput, env, quoted = TRUE)
}
